activerecord/lib/active_record/locking/optimistic.rb

```diff
@@ -80,7 +80,8 @@ def _update_record(attribute_names = self.attribute_names) #:nodoc:
           return 0 if attribute_names.empty?
 
           lock_col = self.class.locking_column
-          previous_lock_value = send(lock_col).to_i
+          previous_lock_value = read_attribute_before_type_cast(lock_col)
```

Comment:
I left this line, but added after it following line:
```
previous_lock_value = nil if previous_lock_value == '0'
```

and removed (overwrote) `_create_record` method above. But in this case we never need to set default value. Just consider it as a possible fix.